// Copyright IBM Corp. All Rights Reserved.
// Copyright 2020 Intel Corporation
//
// SPDX-License-Identifier: Apache-2.0

syntax = "proto3";

import "google/protobuf/any.proto";
// Imports from fabric ..
// - 'protos' package
import "peer/chaincode.proto";
import "peer/proposal.proto";
import "peer/proposal_response.proto";
// - 'kvrwset' package
import "ledger/rwset/kvrwset/kv_rwset.proto";

package fpc;
option go_package = "github.com/hyperledger-labs/fabric-private-chaincode/internal/protos";

message CCParameters {
    // name of the chaincode
    string chaincode_id = 1;

    // version of chaincode, this is the expected mrenclave
    // encoded as hexstring (as generated by common/enclave/generate_mrenclave.sh)
    string version = 2;

    // chaincode sequence number
    int64 sequence = 3;

    // name of channel
    string channel_id = 4;
}

message HostParameters {
    // MSP ID of organization hosting (embracing) the peer with corresponding enclave
    // TODO: rename to `creator_msp_id` because this value will be set to the creator's mspid (not the peer's)
    string peer_msp_id = 1;

    // the (externally accessible) address of the peer endpoint in format <ip-addr|hostname>:<port-number>
    string peer_endpoint = 2;

    // Post-MVP: This will be a X509 peer certificate on
    // Enclave_VK and CCParameters signed by a CA rooted in
    // the MSP of the Organization hosting the peer running the FPC
    // Chaincode enclave.  This shows the "ownership" of Org for that
    //particular FPC Chaincode enclave.   See additional information in
    // fpc-registration.puml in the 'Org-Enclave binding/certification' group.
    // Note that this field may be moved elsewhere.
    bytes certificate = 3;
}

message AttestedData {
    CCParameters cc_params = 1;

    HostParameters host_params = 2;

    // chaincode enclave public key
    bytes enclave_vk = 3;

    // SHA256 hash of the channel genesis block;
    bytes channel_hash = 4;

    // expected TLCC mrenclave
    // encoded as hexstring (as generated by common/enclave/generate_mrenclave.sh)
    string tlcc_mrenclave = 5;
}

message Credentials {
    // serialization of type **AttestedData**
    google.protobuf.Any serialized_attested_data = 1;

    // serialized attestation/quote as output by `get_attestatation`, see `interfaces.attestation.md`
    bytes attestation = 2;

    // serialized attestation evidence as output by `AttestationToEvidence`, see `interfaces.attestation.md`
    bytes evidence = 3;
}

message InitEnclaveMessage {
    // the (externally accessible) address of the peer endpoint in format <ip-addr|hostname>:<port-number>
    string peer_endpoint = 1;

    // opaque parameters passed for initialization of the attestation API
    // (see more in 'attestation.proto', 'interfaces.attestation.md' and 'common/crypto/attestation-api')
    bytes attestation_params = 2;
}

message CleartextChaincodeRequest {
    // the function and args to invoke
    protos.ChaincodeInput input = 1;

    // key to (aes128-gcm) encrypt ChaincodeResponse
    bytes return_encryption_key = 3;
}

message ChaincodeRequestMessage {
    // an RSA-encryption of the serialization of CleartextChaincodeRequest with the chaincode encryption key
    bytes encrypted_request = 1;
}

message CleartextChaincodeResponse {
    // the response of the chaincode invocation
    protos.Response response = 1;
}

message ChaincodeResponseMessage {
    // an aes128-gcm encryption of the serialization of CleartextChaincodeResponse
    bytes encrypted_response = 1;

    // R/W set (of cleartext keys but encrypted values)
    // This field is only valid for the FPC Lite variant but absent from the full version with in-peer FPC validation
    kvrwset.KVRWSet rw_set = 2;

    // signed proposal for this request
    protos.SignedProposal proposal = 3;

    // signature over proposal, cleartext response, (hash of) R/W set
    bytes signature = 4;

    // identity for public key used to sign
    string enclave_id = 5;
}
